cpp/src/arrow/array/data.cc

```diff
@@ -228,22 +229,20 @@ void ArraySpan::SetMembers(const ArrayData& data) {
 namespace {
 
 template <typename offset_type>
-void SetOffsetsForScalar(ArraySpan* span, offset_type* buffer, int64_t value_size,
-                         int buffer_index = 1) {
-  buffer[0] = 0;
-  buffer[1] = static_cast<offset_type>(value_size);
-  span->buffers[buffer_index].data = reinterpret_cast<uint8_t*>(buffer);
-  span->buffers[buffer_index].size = 2 * sizeof(offset_type);
+BufferSpan OffsetsForScalar(uint8_t* scratch_space, offset_type value_size) {
+  auto* offsets = reinterpret_cast<offset_type*>(scratch_space);
+  offsets[0] = 0;
+  offsets[1] = static_cast<offset_type>(value_size);
+  return {scratch_space, sizeof(offset_type) * 2};
 }
 
 int GetNumBuffers(const DataType& type) {
   switch (type.id()) {
     case Type::NA:
     case Type::STRUCT:
     case Type::FIXED_SIZE_LIST:
-      return 1;
     case Type::RUN_END_ENCODED:
-      return 0;
+      return 1;
```

Comment:
I merged a bit quickly, why does `RUN_END_ENCODED` need one buffer here?

Reply:
Many places in the codebase assume that `buffers.size() >= 1`, even if `buffers[0] == nullptr`. When I added test cases which exercised REE scalars those places segfaulted. I thought that requiring `buffers.size() >= 1` for REE (as we do for union) was the most expeditious fix

Reply:
@felipecrv What do you think here? Should we require a one-element `buffers` vector for REE?

Reply:
`RUN_END_ENCODED` doesn't need any buffer but `NA` also does not and we return `1` here. :thinking: 

It's a "once you start lying you can't stop lying" kind of problem for `GetNumBuffers`

Reply:
FWIW, this was the only place in the codebase which didn't already give REE at least one buffer; the constructors, the builder, ... already did so